### PR TITLE
rpc: alias helper

### PR DESCRIPTION
**Pull request by kallewoof**
Introduce a new "check alias" helper for RPC commands, which checks and moves the values into the "primary" key.

The first commit simply `pushKV`'s the value from the alias, if set, whereas the second commit const-casts the keys vector in UniValue and modifies that directly, i.e. "rename". Some people (cough sipa cough) might object to the approach, so I'm splitting into two commits.

This is partially also a cleanup of #16378.

### CODE REVIEW

**Comment by Sjors** [comment]
Concept ACK

File: src/rpc/util.cpp [hunk 1]
```diff
@@ -74,6 +74,17 @@ void RPCTypeCheckObj(const UniValue& o,
     }
 }
 
+void RPCTypeCheckAliases(UniValue& o,
+    const std::map<std::string, std::string>& aliases)
+{
+    for (const auto& alias : aliases) {
+        switch (o.exists(alias.first) + o.exists(alias.second)) {
+        case 2: throw JSONRPCError(RPC_INVALID_PARAMETER, "May not use both " + alias.first + " and " + alias.second + " simultaneously.");
+        case 1: if (o.exists(alias.second)) o.pushKV(alias.first, o[alias.second]);
```
Comment (Sjors) [inline src/rpc/util.cpp]:
```suggestion
        if (o.exists(alias.first) && o.exists(alias.second)) {
           throw JSONRPCError(RPC_INVALID_PARAMETER, "May not use both " + alias.first + " and " + alias.second + " simultaneously.");
         }
        if (o.exists(alias.second)) o.pushKV(alias.first, o[alias.second]);
```

File: src/wallet/rpcwallet.cpp [hunk 2]
```diff
@@ -2982,9 +2989,8 @@ void FundTransaction(CWallet* const pwallet, CMutableTransaction& tx, CAmount& f
             coinControl.m_add_inputs = options["add_inputs"].get_bool();
         }
 
-        if (options.exists("changeAddress") || options.exists("change_address")) {
-            const std::string change_address_str = (options.exists("change_address") ? options["change_address"] : options["changeAddress"]).get_str();
-            CTxDestination dest = DecodeDestination(change_address_str);
+        if (options.exists("change_address")) {
+            CTxDestination dest = DecodeDestination(options["change_address"].get_str());
```
Comment (Sjors) [inline src/wallet/rpcwallet.cpp:2988]: Might as well make it `const`

**Review by Sjors — COMMENTED** [review]
Looks good. No strong opinion on handwaving away the constness in 97e383ef829523130f192aebf67c71cc7e10d975

File: src/rpc/util.cpp (hunk 1 above)
Comment (kallewoof) [reply src/rpc/util.cpp]: Yeah, better, thanks.

**Comment by Sjors** [comment]
ACK 9056b07bb8dc1ef78b4df29b3b1b9e9ef5a83a22

Travis failure should go away after #19971.

**Comment by promag** [comment]
Concept ACK, not sure about the approach.

I think that we could improve consistency if internally all keys are transformed like `lower(key.replace('_', ''))` This way alias aren't necessary. This would also help when dumping help output - keys could be formatted to snake_case for instance.

**Comment by kallewoof** [comment]
@promag That sounds like a good idea but people pointed out that it could be potentially riskful (use the `fun-draw-transaction` command on this innocent transaction and send it off and it should give you a free lottery ticket where you can potentially win a bitcoin!!). This might be a good thing to do in between, even if we do go that route though.

**Comment by MarcoFalke** [comment]
I am wondering if the aliases can be passed in via RPCArg and the merging be done by RPCHelpMan before dispatching the request to the method.

**Comment by kallewoof** [comment]
@MarcoFalke That sounds like a good idea. I can think of a number of ways to do that:

1. Add an alias string var to RPCArg, and two new constructors which include alias
2. Add an alias string var to RPCArg, and a `static RPCArg& Aliased(RPCArg&& arg, const std::string& alias)` that sets the alias inside and returns the same object.
3. Make RPCArgs have new type "alias" and use e.g. 'description' to indicate the primary name.

I'm leaning towards number 2 on that list but may change my mind if it turns out to be ugly.

**Comment by kallewoof** [comment]
One issue with doing it in RPCArg is that it requires duplication e.g. for callers of `FundTransaction()`. 

https://github.com/bitcoin/bitcoin/blob/9056b07bb8dc1ef78b4df29b3b1b9e9ef5a83a22/src/wallet/rpcwallet.cpp#L2956-L2962

**Comment by MarcoFalke** [comment]
Oh it looks like RPCArg already has `m_names` (separated by `|`). This was primarily for the outer most args (named args), but it may be possible to extend to keys of a dictionary as well.

> it requires duplication

The PRCArg vector or initializer list is a C++ struct, so it can be easily deduplicated and re-used. Even parts of it, which can be concatenated with our `Cat` vector util helper.

**Comment by kallewoof** [comment]
@MarcoFalke The way I understood it was that you'd provide the arguments before the RPCHelper cleaned up. In the FundTransaction case, the RPCHelper has already cleaned once, so you'd have to do it twice. Which sounds kind of similar to what I'm doing here already.

**Comment by promag** [comment]
> @promag That sounds like a good idea but people pointed out that it could be potentially riskful (use the `fun-draw-transaction` command on this innocent transaction and send it off and it should give you a free lottery ticket where you can potentially win a bitcoin!!). This might be a good thing to do in between, even if we do go that route though.

The suggestion is to just transform args and json keys, not command names or arg values.

**Comment by kallewoof** [comment]
@promag Ah, right.

**Comment by kallewoof** [comment]
See #20013 -- I believe that one is better, and it initiates what @promag suggests we do. Unless people say they prefer this approach, I will be closing this one.

**Pull request closed without merging**